code/controllers/subsystem/processing/quirks.dm

```diff
@@ -65,3 +65,74 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
 	// Assign wayfinding pinpointer granting quirk if they're new
 	if(cli.get_exp_living(TRUE) < EXP_ASSIGN_WAYFINDER && !user.has_quirk(/datum/quirk/needswayfinder))
 		user.add_quirk(/datum/quirk/needswayfinder, TRUE)
+
+/datum/controller/subsystem/processing/quirks/proc/randomise_quirks(mob/living/user, bonus_points = 0)
+	var/bonus_quirks = max((length(user.roundstart_quirks) + rand(-3,3)), 3)
+	var/added_quirk_count = 0 //How many we've added
+	var/list/quirks_to_add = list() //Quirks we're adding
+	var/loop_preventer = 10 // Makes sure we don't get stuck
+	var/good_count = 0 //Maximum of 6 good perks
+	var/score //What point score we're at
+	//Create a random list of stuff to start with
+	while(bonus_quirks > added_quirk_count)
+		if(!loop_preventer)//Lets not get stuck
+			break
+		var/quirk = pick(quirks) //quirk is a string
+		if((quirk in quirk_blacklist) && (quirk in quirks_to_add)) //prevent blacklisted
+			loop_preventer--
+			continue
+		if(quirk in quirks_to_add)
+			loop_preventer--
+			continue
+		if(quirk_points[quirk] > 0)
+			good_count++
+		score += quirk_points[quirk]
+		quirks_to_add += quirk
+		added_quirk_count++
```

Comment:
This shouldn't be possible to infinite loop.

Copy the quirks list, and just remove whatever is blacklisted and used. That way, you'll only ever get valid entries.

Reply:
Here and in the other place.

Reply:
I tightened up the loops and did as you said!